xds/src/main/java/io/grpc/xds/ClientXdsClient.java

```diff
@@ -228,86 +238,259 @@ private static LdsUpdate processClientSideListener(Listener listener, boolean pa
+      throw new ResourceInvalidException(
+          "Listener " + proto.getName() + " cannot have listener_filters");
+    }
+    if (proto.hasUseOriginalDst()) {
+      throw new ResourceInvalidException(
+          "Listener " + proto.getName() + " cannot have use_original_dst set to true");
+    }
+
+    String address = null;
+    if (proto.getAddress().hasSocketAddress()) {
+      SocketAddress socketAddress = proto.getAddress().getSocketAddress();
+      address = socketAddress.getAddress();
+      switch (socketAddress.getPortSpecifierCase()) {
+        case NAMED_PORT:
+          address = address + ":" + socketAddress.getNamedPort();
+          break;
+        case PORT_VALUE:
+          address = address + ":" + socketAddress.getPortValue();
+          break;
+        default:
+          // noop
+      }
+    }
+
+    List<FilterChain> filterChains = new ArrayList<>();
+    for (io.envoyproxy.envoy.config.listener.v3.FilterChain fc : proto.getFilterChainsList()) {
+      filterChains.add(
+          parseFilterChain(fc, rdsResources, tlsContextManager, filterRegistry, parseHttpFilter));
+    }
+    FilterChain defaultFilterChain = null;
+    if (proto.hasDefaultFilterChain()) {
+      defaultFilterChain = parseFilterChain(
+          proto.getDefaultFilterChain(), rdsResources, tlsContextManager, filterRegistry,
+          parseHttpFilter);
+    }
+
+    return new EnvoyServerProtoData.Listener(
+        proto.getName(), address, Collections.unmodifiableList(filterChains), defaultFilterChain);
+  }
+
+  @VisibleForTesting
+  static FilterChain parseFilterChain(
+      io.envoyproxy.envoy.config.listener.v3.FilterChain proto, Set<String> rdsResources,
+      TlsContextManager tlsContextManager, FilterRegistry filterRegistry, boolean parseHttpFilters)
+      throws ResourceInvalidException {
+    io.grpc.xds.HttpConnectionManager httpConnectionManager = null;
+    HashSet<String> uniqueNames = new HashSet<>();
+    for (io.envoyproxy.envoy.config.listener.v3.Filter filter : proto.getFiltersList()) {
+      if (!uniqueNames.add(filter.getName())) {
+        throw new ResourceInvalidException(
+            "FilterChain " + proto.getName() + " with duplicated filter: " + filter.getName());
+      }
+      if (!filter.hasTypedConfig()) {
+        throw new ResourceInvalidException(
+            "FilterChain " + proto.getName() + " contains filter " + filter.getName()
+                + " without typed_config");
+      }
+      Any any = filter.getTypedConfig();
+      // HttpConnectionManager is the only supported network filter at the moment.
+      if (!any.getTypeUrl().equals(TYPE_URL_HTTP_CONNECTION_MANAGER)) {
+        throw new ResourceInvalidException(
+            "FilterChain " + proto.getName() + " contains filter " + filter.getName()
+                + " with unsupported typed_config type " + any.getTypeUrl());
+      }
+      // Any filters after the first HttpConnectionManager are ignored.
```

Comment:
The code doesn't actually ignore filters after the first HCM. If you break the loop after line 335 it would have.

Reply:
Why? there is a `if (httpConnectionManager == null)` statement guarding the code of parsing HCM. We still want to loop over the list of network filters, just in case there are filters with duplicated name or filters other than HCM.

Reply:
> ... We still want to loop over the list of network filters, just in case there are filters with duplicated name or filters other than HCM.

That is *not* ignoring and the comment is wrong. But if the code matches the spec (from A36 gRFC) then it's a minor issue. The spec tells to validate all of them but not use any for processing. So the code seems to be right (but the comment doesn't match)

Reply:
Well, the comment is really intend to mean we will take the first HttpConnectionManager. I will update or delete the comment if you think that is confusing.

Reply:
Deleted the comment.